### ISSUE TITLE
please include sh or bash in the image

### ISSUE THREAD

**Issue by eduarddrenth** [eXist-db/docker-existdb#72]
since there is no sh or bash I cannot interact with a running container which makes docker development very hard

**Comment by dizzzz** [comment]
please could you rephrase your ticket more as question or concrete suggestion? It is more like a one-liner now, not stating what you think what should be changed. That would be a start of a clear discussion.

AFAIK the docker image is made as small as possible on purpose; a shell could un-harden the image even?

**Comment by eduarddrenth** [comment]
For interacting with a container it would be very helpfull to have sh and/or bash available as well as some basic shell commands

**Comment by eduarddrenth** [comment]
I personally think a minimal shell is a must. For example payara and postgres images also include that.

**Comment by duncdrum** [comment]
as mentioned in the [readme](https://github.com/eXist-db/docker-existdb#interacting-with-the-running-container) these images are based on distroless images the exclusion of a linux shell is very much by design. 

To interact with the running container issue `docker exec …` commands or use the Java Admin Client to issue shell commands. If you wish to attach your local shell more permanently you can enable the `:debug` mode of the distroless image [here](https://github.com/eXist-db/docker-existdb/blob/11627482b2f864093fccc311f57c813a07d8a65f/Dockerfile#L148). This step, however, requires you to work with your own base images, as the official images are designed to be used in production as base images, we will not ship with a `:debug` base images for the reasons that Dannes already mentioned.

**Comment by grantmacken** [comment]
Most of my shell interaction with the eXist db running in a container,
is achieved thru eXist rest api. 
I mostly use the pattern outlined in following bash script 

```
#!/usr/bin/env bash
IPAddress=$(docker inspect --format='{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}'  exist)
URL="http://${IPAddress}:8080/exist/rest/db"

function post() {
curl  -s \
  -H 'Content-Type: application/xml' \
  -u 'admin:' \
  --data-binary @- "$URL"
}

cat <<EOF | post
<query xmlns='http://exist.sourceforge.net/NS/exist' start='1' max='99' wrap='no'>
<text><![CDATA[
xquery version '3.1';
repo:list() =>
 string-join('&#10;')
]]></text>
</query>
EOF
```
If you need to interact with the file system on the container,
the file module allows you to list move, make directories etc

**Comment by eduarddrenth** [comment]
ok, the reason for me to ask this is I want to get an image with xar's that are already deployed and indexed. OS access would be very welcome here, for example to start and stop exist.

**Comment by duncdrum** [comment]
restarting the container will reboot exist, for reindexing you can use either xquery commands via `docker exec` or you can modify the still image before starting the container.

**Comment by eduarddrenth** [comment]
I know these options yes, but I want one Dockerfile to prepare a ready to use image

**Comment by eduarddrenth** [comment]
I guess I'll persist using docker volumes

**Comment by duncdrum** [comment]
I m confused now, to start and stop exist from within a `Dockerfile` use multiple `FROM`, you can prepare indexing operations using `RUN` as documented in the readme. 

If something isn't working please open a new issue with steps to reproduce, as none of these operations seems to depend on the existence or lack of shell.

**Comment by eduarddrenth** [comment]
I now about those options yes, but find it easier the way I have it now, see: https://bitbucket.org/fryske-akademy/exist-db/src/master/docker/

**Comment by duncdrum** [comment]
yup looks good, docker-compose is certainly the elegant solution, also recommended in the readme.

`docker exec -it <container> "/bin/bash"` will never work it should be:

`docker exec <container> java -jar start.jar client` to talk to exist, or a variation thereof to talk to either jetty or the JVM

**Comment by bptarpley** [comment]
Would it be possible to provide more examples in the README for how to do DevOps oriented things like listing the file contents of a specific directory using docker exec commands? When a client asks for data to persist in an eXistdb container, it's difficult for sysadmins to know exactly what directory to copy from the running container using "docker cp" commands (so they can then mount it afterward) without being able to poke around inside the container. How, for instance, might one list the contents of the /exist/webapp directory? I perused the documentation for the Java admin client (https://exist-db.org/exist/apps/doc/java-admin-client.xml#command-line) and it seems more geared toward developers who wish to perform maintenance tasks on the database, not the people who are tasked with actually running and maintaining the containers. Many thanks in advance for considering this.

**Comment by line-o** [comment]
to interact with docker instances you can also use [node-exist](https://github.com/eXist-db/node-exist) and the [cli-tools](https://github.com/eXist-db/node-exist#command-line-scripts) it exposes.
